### PR TITLE
drivers: watchdog: wdt_nrfx.c: Unistall timeouts in wdt_disable()

### PR DESCRIPTION
**Pull request by nordic-segl**
According to API description, wdt_disable() shall
automatically uninstall all timeouts.
https://docs.zephyrproject.org/latest/hardware/peripherals/watchdog.html

Implement missing wdt_disable() functionality:
- disable WDT RR channels;
- remove callback from timeout channels;
- set data->m_allocated_channels to zero.

### CODE REVIEW

**Comment by nordic-segl** [comment]
DNM before hal_nordic supports WDT channel deallocation.

**Comment by kl-cruz** [comment]
Is it now ready to work on?

**Comment by nordic-segl** [comment]
> Is it now ready to work on?

No,
void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance);
is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?

**Comment by ArekBalysNordic** [comment]
> > Is it now ready to work on?
> 
> No, void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance); is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?

If it is possible it would be great because I would like to have a fix for that as soon as possible. :)

**Comment by nordic-segl** [comment]
> > > Is it now ready to work on?
> > 
> > 
> > No, void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance); is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?
> 
> If it is possible it would be great because I would like to have a fix for that as soon as possible. :)

It's already committed to the hal_nordic repo
https://github.com/zephyrproject-rtos/hal_nordic/commit/8eae3cc15977e4a232e3e5ca1088c048d4ceb263#diff-4e564f3f3dea92d458fc3a9d7c3c2bf6c3b0a6fd2ae9b6906476f6912b03fd10

Here is PR that brings in updated hal_nordic:
https://github.com/zephyrproject-rtos/zephyr/pull/72586

**Review by carlescufi — DISMISSED** [review]

**Review by anangl — DISMISSED** [review]

**Review by masz-nordic — DISMISSED** [review]

File: drivers/watchdog/wdt_nrfx.c [hunk 1]
```diff
@@ -69,6 +71,13 @@ static int wdt_nrf_disable(const struct device *dev)
 		return -EFAULT;
 	}
 
+	nrfx_wdt_channels_free(&config->wdt);
+
+	for (channel_id = 0; channel_id < data->m_allocated_channels; channel_id++) {
+		data->m_callbacks[channel_id] = NULL;
+	}
+	data->m_allocated_channels = 0;
+
```
[the remark this answers is not in the document]
Comment (nordic-segl) [reply drivers/watchdog/wdt_nrfx.c:80]: I add one more test to check if this left over can lead to WDT start without any timeouts configured.

**Review by carlescufi — APPROVED** [review]

File: drivers/watchdog/wdt_nrfx.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nordic-segl) [reply drivers/watchdog/wdt_nrfx.c:80]: I run "extended" wdt_error_cases manually on nrf54l15pdk and nrf54h20. All tests are passing (so fix in this PR is working as expected).

**Review by masz-nordic — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**